cloudify_ansible/utils.py

```diff
@@ -597,3 +626,358 @@ def is_connected_to_internet():
+    try:
+        return json.loads(selected.group(1))
+    except (AttributeError, IndexError):
+        ctx.logger.info('Unable to parse result. '
+                        'Not storing result in runtime properties.')
+        return string_without_whitespace
+
+
+def get_facts(string):
+    """Facts are collected in the ansible check command.
+    This output is returned in string, and then we need to split the string,
+    and then parse the JSON embedded in the string.
+    There is currently not a better way to do this. A thousand curses on
+    Ansible for dumping mixed format output.
+
+    :param string:
+    :return:
+    """
+    new_string = []
+    for line in string.split('\n'):
+        if line.startswith('META'):
+            continue
+        new_string.append(line)
+    separater = '=>(.*)'
+    string_without_whitespace = '\n'.join(
+        new_string).replace("\n", "").replace("\t", "").replace(" ", "")
+    selected = re.search(separater, string_without_whitespace)
+    try:
+        return json.loads(selected.group(1))
+    except (AttributeError, IndexError):
+        ctx.logger.error('Unable to parse result. '
+                         'Not storing result in runtime properties.')
+        return selected
+
+
+def get_tasks_by_host(plays):
+    """Find all the task names associated with a host. Currently not used."""
+    task_names = []
+    for play in plays:
+        for task in play['tasks']:
+            task_names.append(task['task']['name'])
+    return task_names
+
+
+def process_block_tags(block, tasks, tags):
+    """ Get tags from blocks in a playbook. Uses Ansible as a programming
+    library. Copied from Ansible.
+
+    :param block:
+    :param tasks:
+    :param tags:
+    :return:
+    """
+    for task in block.block:
+        if isinstance(task, type(block)):
+            process_block_tags(task, tasks)
+        else:
+            if task.action == 'meta':
+                continue
+            tasks.append(task.name.replace(' ', ''))
+            for tag in task.tags:
+                if tag not in tags:
+                    tags.append(tag)
+
+
+def get_tags_from_playbook(playbook):
+    """ Get a list of tag names
+
+    :param playbook: ansible.playbook.Playbook
+    :return: list
+    """
+    tasks = []
+    tags = []
+    for play in playbook.get_plays():
+        for block in play.compile():
+            process_block_tags(block, tasks, tags)
+    return tasks, tags
+
+
+def get_playbook_from_path(playbook_path):
+    """ Create ansible.playbook.Playbook object.
+
+    :param playbook_path: string
+    :return:
+    """
+    loader = DataLoader()
+    variable_manager = VariableManager(loader=loader)
+    return Playbook.load(
+        file_name=playbook_path,
+        variable_manager=variable_manager,
+        loader=loader)
+
+
+def get_available_steps(playbook_path):
+    return get_tags_from_playbook(get_playbook_from_path(playbook_path))
+
+
+def get_our_tags(all_tags, op_number, max_ops):
```

Comment:
Can you give an example of the invocation of this function? cant understand 100% its purpose